cmd/xl-storage-format-v2.go

```diff
@@ -656,82 +668,82 @@ func (z xlMetaV2) ToFileInfo(volume, path, versionID string) (fi FileInfo, err e
+				return FileInfo{}, errFileNotFound
+			}
+			return FileInfo{}, errFileVersionNotFound
+
 		}
 	}
 
+	orderedVersions := make([]xlMetaV2Version, len(z.Versions))
+	copy(orderedVersions, z.Versions)
+
+	sort.Slice(orderedVersions, func(i, j int) bool {
+		mtime1 := getModTimeFromVersion(orderedVersions[i])
+		mtime2 := getModTimeFromVersion(orderedVersions[j])
+		return mtime1.After(mtime2)
+	})
+
 	if versionID == "" {
-		if len(z.Versions) >= 1 {
-			if !z.Versions[latestIndex].Valid() {
-				logger.LogIf(GlobalContext, fmt.Errorf("invalid version detected %#v", z.Versions[latestIndex]))
-				return FileInfo{}, errFileNotFound
-			}
-			switch z.Versions[latestIndex].Type {
+		if len(orderedVersions) >= 1 {
+			switch orderedVersions[0].Type {
 			case ObjectType:
-				fi, err = z.Versions[latestIndex].ObjectV2.ToFileInfo(volume, path)
-				fi.IsLatest = true
-				return fi, err
+				fi, err = orderedVersions[0].ObjectV2.ToFileInfo(volume, path)
 			case DeleteType:
-				fi, err = z.Versions[latestIndex].DeleteMarker.ToFileInfo(volume, path)
-				fi.IsLatest = true
-				return fi, err
+				fi, err = orderedVersions[0].DeleteMarker.ToFileInfo(volume, path)
 			case LegacyType:
-				fi, err = z.Versions[latestIndex].ObjectV1.ToFileInfo(volume, path)
-				fi.IsLatest = true
-				return fi, err
+				fi, err = orderedVersions[0].ObjectV1.ToFileInfo(volume, path)
 			}
+			fi.IsLatest = true
+			fi.NumVersions = len(orderedVersions)
+			return fi, err
+
 		}
 		return FileInfo{}, errFileNotFound
 	}
 
-	for _, version := range z.Versions {
-		if !version.Valid() {
-			logger.LogIf(GlobalContext, fmt.Errorf("invalid version detected %#v", version))
-			if versionID == "" {
-				return FileInfo{}, errFileNotFound
-			}
-			return FileInfo{}, errFileVersionNotFound
-		}
+	var i = -1
+	var version xlMetaV2Version
+
+findVersion:
+	for i, version = range orderedVersions {
 		switch version.Type {
 		case ObjectType:
 			if bytes.Equal(version.ObjectV2.VersionID[:], uv[:]) {
 				fi, err = version.ObjectV2.ToFileInfo(volume, path)
-				fi.IsLatest = latestModTime.Equal(fi.ModTime)
-				return fi, err
+				break findVersion
 			}
 		case LegacyType:
 			if version.ObjectV1.VersionID == versionID {
 				fi, err = version.ObjectV1.ToFileInfo(volume, path)
-				fi.IsLatest = latestModTime.Equal(fi.ModTime)
-				return fi, err
+				break findVersion
 			}
 		case DeleteType:
 			if bytes.Equal(version.DeleteMarker.VersionID[:], uv[:]) {
 				fi, err = version.DeleteMarker.ToFileInfo(volume, path)
-				fi.IsLatest = latestModTime.Equal(fi.ModTime)
-				return fi, err
+				break findVersion
 			}
 		}
 	}
 
+	if err != nil {
+		return fi, err
+	}
+
+	if i >= 0 {
+		// A version is found, fill dynamic fields
+		fi.IsLatest = i == 0
+		fi.NumVersions = len(z.Versions)
+		if i < len(orderedVersions)-1 {
+			fi.SuccessorModTime = getModTimeFromVersion(orderedVersions[i+1])
+		}
+		return fi, nil
+	}
+
 	if versionID == "" {
 		return FileInfo{}, errFileNotFound
```

Comment:
This is a critical function some tests would be nice to capture the previous behavior as well as new changes

Reply:
PR merged @vadmeste you send the tests in another PR along with the healing skip that we have to do.